File: cuesubmit/cuesubmit/ui/Frame.py
#  Copyright Contributors to the OpenCue Project
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.


"""Widget for entering a frame spec."""


from __future__ import print_function
from __future__ import division
from __future__ import absolute_import

from cuesubmit.ui import Widgets


class FrameSpecWidget(Widgets.CueHelpWidget):
    """Widget for entering a frame spec."""

    helpText = 'Enter a FrameSpec value.\n' \
               'A frame spec consists of a start time, an optional end time, a step, ' \
               'and an interleave.\n' \
               'Multiple ranges can be added together by separating with commas.\n' \
               '    Ex:\n' \
               '    1-10x3\n' \
               '    1-10y3 // inverted step\n' \
               '    10-1x-1\n' \
               '    1 // same as "1-1x1"\n' \
               '    1-10:5 // interleave of 5\n' \
               '    1-5x2, 6-10 // 1 through 5 with a step of 2 and 6 through 10\n'

    def __init__(self, parent=None):
        super(FrameSpecWidget, self).__init__(parent)
        self.frameSpecInput = Widgets.CueLabelLineEdit('Frame Spec:')
        self.contentLayout.addWidget(self.frameSpecInput)

File: rust/crates/scheduler/src/models/frame.rs
// Copyright Contributors to the OpenCue Project
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License. You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied. See the License for the specific language governing permissions and limitations under
// the License.

use std::{collections::HashMap, fmt::Display, time::SystemTime};

use bytesize::ByteSize;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::models::{core_size::CoreSize, fmt_uuid};

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct DispatchFrame {
    // Entity fields
    pub id: Uuid,
    pub frame_name: String,

    // LayerEntity fields
    pub show_id: Uuid,
    pub facility_id: Uuid,
    pub job_id: Uuid,

    // FrameEntity fields
    pub layer_id: Uuid,

    // DispatchFrame specific fields
    pub command: String,
    pub range: String,
    pub chunk_size: i32,
    pub show_name: String,
    pub shot: String,
    pub user: String,
    pub uid: Option<i32>,
    pub log_dir: String,
    pub layer_name: String,
    pub job_name: String,
    // Min cores can be a negative, representing `machine_total_cores - min_cores`
    pub min_cores: CoreSize,
    pub layer_cores_limit: Option<CoreSize>,
    pub threadable: bool,
    pub has_selfish_service: bool,
    pub min_gpus: u32,
    pub min_gpu_memory: ByteSize,
    pub min_memory: ByteSize,
    pub slots_required: u32,
    // On Cuebot these fields come from constants, maybe replicate these constants here
    // pub int_soft_memory_limit: i64,
    // pub int_hard_memory_limit: i64,
    pub services: Option<String>,
    pub os: Option<String>,
    pub loki_url: Option<String>,
    pub version: u32,
    pub updated_at: SystemTime,
    pub env: HashMap<String, String>,
}

impl Display for DispatchFrame {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}.{}.{}({})",
            self.job_name,
            self.layer_name,
            self.frame_name,
            fmt_uuid(&self.id)
        )
    }
}
